### ISSUE TITLE
Enhancement Proposal: Multi-Camera Feature Fusion with Batch-Aware Architecture

### ISSUE THREAD

**Issue by riita10069** [autowarefoundation/auto_e2e#2]
## Summary

The current `FeatureFusion` module does not perform cross-camera fusion — it only concatenates multi-scale features from a single image along the channel dimension. Additionally, the `DrivingPolicy` module uses `torch.flatten()` without preserving the batch dimension, making mini-batch training impossible. This proposal outlines three approaches to properly unify multi-camera inputs while supporting batched training/inference.

### 1. No cross-camera fusion
`FeatureFusion.forward()` receives multi-scale features with shape `[B*V, C, H, W]` (where V=num_views) and concatenates scales along channels. Each camera's features remain independent — there is no mechanism to reason about spatial relationships across camera views.

### 2. Batch dimension collapse
`DrivingPolicy.forward()` calls `torch.flatten(feature_map)` which collapses all dimensions including batch into a single 1D vector. The Linear layer input sizes (2328) are hardcoded to match exactly `batch=8` (i.e., num_views=8), making it impossible to train with batch_size > 1.

### 3. FutureState outputs per-camera predictions
Since camera features are never unified, `FutureState` predicts future visual features independently per camera rather than for the unified scene.

---

### Approach A: Parametrized View Reshaping (Minimal Change)

Parametrize the model with `num_views` and use reshape/split to handle the view dimension explicitly:

```python
class AutoE2E(nn.Module):
    def __init__(self, num_views=8):
        super().__init__()
        self.num_views = num_views
        ...

    def forward(self, x, visual_history, egomotion_history):
        # x: [B, V, C, H, W]
        B, V, C, H, W = x.shape
        
        # Merge batch and views for backbone processing
        x = x.reshape(B * V, C, H, W)
        features = self.Backbone(x)
        
        # Fuse multi-scale features, then reunify views
        fused = self.FeatureFusion(features)  # [B*V, 1440, 7, 7]
        fused = fused.reshape(B, V * 1440, 7, 7)  # Merge view channels
        
        # Reduce to scene-level representation
        fused = self.view_reduce(fused)  # Conv2d(V*1440, 1440, 1) → [B, 1440, 7, 7]
        ...
```

**Pros:** Minimal architectural change, fixes both the batch dimension bug and enables configurable num_views.  
**Cons:** Simple channel concatenation + Conv does not model spatial relationships between cameras.

---

### Approach B: Cross-Camera Spatial Attention

Apply a transformer-style attention mechanism across camera views to learn spatial relationships, similar to how PETR/PETRv2 encode 3D position information into 2D features.

```python
class CrossViewAttention(nn.Module):
    def __init__(self, embed_dim=1440, num_heads=8, num_views=8):
        super().__init__()
        self.camera_embed = nn.Embedding(num_views, embed_dim)
        self.attn = nn.MultiheadAttention(embed_dim, num_heads, batch_first=True)
    
    def forward(self, fused_features, B, V):
        # fused_features: [B*V, C, H, W]
        C, H, W = fused_features.shape[1:]
        x = fused_features.reshape(B, V, C, H * W)  # [B, V, C, HW]
        x = x.permute(0, 3, 1, 2).reshape(B * H * W, V, C)  # [B*HW, V, C]
        
        # Add learnable camera position embeddings
        cam_pos = self.camera_embed.weight.unsqueeze(0)  # [1, V, C]
        x = x + cam_pos
        
        # Cross-view attention
        x, _ = self.attn(x, x, x)  # [B*HW, V, C]
        
        # Pool across views
        x = x.mean(dim=1)  # [B*HW, C]
        return x.reshape(B, H, W, C).permute(0, 3, 1, 2)  # [B, C, H, W]
```

**Pros:** Learns which cameras are relevant to each other at each spatial location. Aligned with state-of-the-art approaches (PETR, UniAD).  
**Cons:** Higher compute cost, requires careful camera position encoding design.

**Related Work:**
- PETR [[1]](#references) encodes 3D coordinates into 2D features via position embedding transformation
- PETRv2 [[2]](#references) extends this with temporal alignment across frames
- UniAD [[3]](#references) uses cross-attention query design across all driving tasks
- Tesla's HydraNet [[6]](#references) processes all camera inputs through a shared backbone with cross-camera attention in the BEV transformer stage

---

### Approach C: Bird's Eye View (BEV) Projection

Project per-camera features into a shared BEV coordinate frame using camera intrinsic/extrinsic parameters, then perform planning in BEV space.

```
Per-camera image → Backbone → 2D Features
                                    ↓
              Depth prediction (per pixel) or spatial cross-attention
                                    ↓
                   BEV Feature Grid: [B, C, X, Y]
                                    ↓
                   DrivingPolicy / FutureState
```

Two sub-approaches:
1. **Explicit depth (LSS-style):** Predict per-pixel depth distribution, "lift" features to 3D voxels, "splat" onto BEV grid.
2. **Implicit projection (BEVFormer-style):** Use deformable attention with 3D reference points to query image features from BEV grid positions.

**Pros:** Geometrically principled, enables metric-space reasoning, proven in production (Tesla Occupancy Networks). Natural representation for trajectory planning.  
**Cons:** Requires camera calibration parameters as input. Higher memory footprint. More complex implementation.

**Related Work:**
- LSS [[4]](#references) introduced the lift-splat-shoot paradigm
- BEVFormer [[5]](#references) uses spatiotemporal transformers to generate BEV from multi-camera inputs
- VAD [[7]](#references) performs vectorized planning in BEV space end-to-end
- Tesla Occupancy Networks [[6]](#references) extend BEV to full 3D volumetric occupancy prediction from cameras only

---

## Recommended Path Forward

| Phase | Approach | Goal |
|-------|----------|------|
| **Phase 1** | Approach A | Fix batch dimension bug, parametrize num_views, unblock training loop development |
| **Phase 2** | Approach B | Add cross-camera attention for meaningful multi-view fusion |
| **Phase 3** | Approach C | Migrate to BEV representation for geometric reasoning and trajectory planning |

Phase 1 is a prerequisite for any training to happen. Phases 2 and 3 can be evaluated against each other once a training loop and dataset are in place.

---

## References

1. Liu, Y., Wang, T., Zhang, X., et al. "PETR: Position Embedding Transformation for Multi-View 3D Object Detection." ECCV 2022. https://arxiv.org/abs/2203.05625
2. Liu, Y., Yan, J., et al. "PETRv2: A Unified Framework for 3D Perception from Multi-Camera Images." ICCV 2023. https://arxiv.org/abs/2206.01256
3. Hu, Y., Yang, J., Chen, L., et al. "Planning-oriented Autonomous Driving." CVPR 2023 (Best Paper). https://arxiv.org/abs/2212.10156
4. Philion, J., Fidler, S. "Lift, Splat, Shoot: Encoding Images From Arbitrary Camera Rigs by Implicitly Unprojecting to 3D." ECCV 2020. https://arxiv.org/abs/2008.05711
5. Li, Z., Wang, W., Li, H., et al. "BEVFormer: Learning Bird's-Eye-View Representation from Multi-Camera Images via Spatiotemporal Transformers." ECCV 2022. https://arxiv.org/abs/2203.17270
6. Tesla, Inc. "Tesla AI Day 2022." Technical presentation, September 30, 2022. https://www.youtube.com/watch?v=ODSJsviD_SU
7. Jiang, B., Chen, S., Xu, Q., et al. "VAD: Vectorized Scene Representation for Efficient Autonomous Driving." ICCV 2023. https://arxiv.org/abs/2303.12077

**Comment by m-zain-khawaja** [comment]
@riita10069 - these are excellent suggestions.

1. Could you please submit a PR for paramaterization of the inputs so that we can do larger mini-batch size training.
2. I like the idea of cross-camera spatial attention, could you please submit a PR which includes this module in the network.
3. BEV projection is also a great suggestion. I suggest that we can leverage the spatial cross attention mechanism instead of depth prediction. Could you also please submit a PR with this module.


Thank you!

**Comment by m-zain-khawaja** [comment]
@riita10069 - I have merged the parameterization and cross-camera spatial attention module you submitted, thank you for this contribution. Are you able to also submit a PR for BEV projection module.

**Comment by m-zain-khawaja** [comment]
Thanks @riita10069
